### PR TITLE
docs: add valid chars in image name and tag 

### PR DESCRIPTION
**Pull request by isubuz**
Documented the valid characters in an image name or tag in the tag reference section. Added links to the tag reference page from other pages which refers to an image name or tag. Addresses #20788 

Verified that the links work in the generated docs.

This is just the documentation change. The remaining code changes to update the error message will be done in another PR.

### CODE REVIEW

**Comment by aaronlehmann** [comment]
ping @dmcgowan @stevvooe

File: docs/reference/commandline/tag.md [hunk 1]
```diff
@@ -16,5 +16,11 @@ parent = "smn_cli"
 
       --help               Print usage
 
+Image name must match the regular expression `[a-z0-9]+`.
```
Comment (aaronlehmann) [inline docs/reference/commandline/tag.md]: The rules for image names are actually a bit more restrictive. Here's what I wrote up in #22264:

> A repository name is made up of slash-separated
> name components, optionally prefixed by a DNS hostname. The hostname
> must follow comply with standard DNS rules, but may not contain
> <code>_</code> characters. If a hostname is present, it may optionally
> be followed by a port number in the format <code>:8080</code>.
> Name components may contain lowercase characters, digits, and
> separators. A separator is defined as a period, one or two underscores,
> or one or more dashes. A name component may not start or end with
> a separator.

File: man/docker-commit.1.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ CONTAINER [REPOSITORY[:TAG]]
 Create a new image from an existing container specified by name or
 container ID.  The new image will contain the contents of the
 container filesystem, *excluding* any data volumes.
+[Read more about valid image names and tags](https://docs.docker.com/reference/commandline/tag/).
```
Comment (aaronlehmann) [inline man/docker-commit.1.md]: I'm not sure these links will show up once these files are postprocessed to create manual pages. Does anyone know how this is handled?

Comment (thaJeztah) [inline man/docker-commit.1.md]: It may be better to refer to the "docker tag" man page, instead of a link to the documentation. People using the man pages usually don't want to open a browser for this, for example;

```Markdown
Refer to **docker-tag(1)** for more information
about valid image- and tag names.
```

File: docs/reference/commandline/tag.md (hunk 1 above)
Comment (stevvooe) [inline docs/reference/commandline/tag.md]: Also, for reference, here is _actual_ [grammar](https://github.com/docker/distribution/blob/master/reference/reference.go#L4).

File: docs/reference/commandline/tag.md [hunk 3]
```diff
@@ -16,5 +16,11 @@ parent = "smn_cli"
 
       --help               Print usage
 
+Image name must match the regular expression `[a-z0-9]+`. 
+
+Image tag must match the regular expression `[\w][\w.-]{0,127}` where `\w`
```
Comment (stevvooe) [inline docs/reference/commandline/tag.md]: It is probably okay to shorten this to have the expanded character class, as `\w` hides the combination.

Double check, but I think `[A-Za-z0-9_][A-Za-z0-9_-.]` is correct.

File: man/docker-commit.1.md (hunk 2 above)
Comment (isubuz) [inline man/docker-commit.1.md]: Makes sense.

File: docs/reference/commandline/tag.md (hunk 1 above)
Comment (isubuz) [inline docs/reference/commandline/tag.md]: The tag reference currently documents the command as:

```docker tag [OPTIONS] IMAGE[:TAG] [REGISTRYHOST/][USERNAME/]NAME[:TAG]```

Why do we specifically mention `USERNAME` component in the 4th argument and not in the 3rd? Unless needed for specific reasons, IMHO, it is a bit confusing as username is part of the image name. Shouldn't it be:

```docker tag [OPTIONS] [USERNAME/]NAME[:TAG] [REGISTRYHOST/][USERNAME/]NAME[:TAG]```

or

```docker tag [OPTIONS] IMAGE[:TAG] [REGISTRYHOST/]IMAGE[:TAG]```

I would vote for the first as it more explicit and clear. I can then document the regexes for all components.

On a side note, `REGISTRYHOST` can be specified in the 3rd argument as well. Note sure if this should be flagged as a CLI parser error.

Comment (thaJeztah) [inline docs/reference/commandline/tag.md]: I think that may be from a "best practice" perspective. It's generally discouraged to tag your local images with the "username/name:tag" format, because doing so would overwrite your local images when you `docker pull` that same image from the repository. Omitting `username/` for your local images makes it easier to distinguish your locally built images from the ones you pulled.

(Just my thoughts why it was written that way)

Comment (isubuz) [inline docs/reference/commandline/tag.md]: Makes sense. In that case, should the command be documented as:

```docker tag [OPTIONS] NAME[:TAG] [REGISTRYHOST/][USERNAME/]NAME[:TAG]```

Note `NAME` instead of `IMAGE`. To me it seems like we want to refer to the name component only.

Comment (thaJeztah) [inline docs/reference/commandline/tag.md]: Hm, I think that's correct, yes.

(but @stevvooe or @aaronlehmann are the experts in conventions for these names :smile:)

Comment (aaronlehmann) [inline docs/reference/commandline/tag.md]: I think it's better not to include `[USERNAME/]`, because this isn't actually part of the reference grammar; it's just convention. It could be confusing for these docs to diverge from the actual grammar.

Comment (stevvooe) [inline docs/reference/commandline/tag.md]: I'd avoid any mention of `REGISTRYHOST` or `USERNAME`. These are interpretive conventions. https://github.com/docker/docker/pull/22665#discussion_r63022941 has the right format, slightly modified to this:

```
docker tag [OPTIONS] NAME[:TAG] NAME[:TAG]
```

Further documentation can clarify that the name may include a registry host, per this aspect of the grammar:

```
name                            := [hostname '/'] component ['/' component]*
```

As an aside, it seems like we should take the time to document that grammar directly. It should be broken down into the various roles:

1. Taggable names (or better word). These can be taken as user input and is the topic of discussion here.
2. The superset, known as references, including digests.

This is probably outside the scope of this PR.

Comment (isubuz) [inline docs/reference/commandline/tag.md]: Pushed the changes. I can work on updating the grammar docs in a separate PR.

**Comment by aaronlehmann** [comment]
Content LGTM.

**Comment by stevvooe** [comment]
Content LGTM

File: docs/reference/commandline/tag.md [hunk 4]
```diff
@@ -10,11 +10,23 @@ parent = "smn_cli"
...
+may not contain underscores. If a hostname is present, it may optionally be
+followed by a port number in the format `:8080`. Name components may contain
+lowercase characters, digits and separators. A separator is defined as a period,
+one or two underscores, or one or more dashes. A name component may not start or
+end with a separator.
+
+A tag name may contain lowercase and uppercase characters, digits, underscores,
+periods and dashes. A tag name may not start with a period or a dash and may
+contain a maximum of 128 characters.
+
```
Comment (thaJeztah) [inline docs/reference/commandline/tag.md]: This is "technically" correct now, but I think having some examples would clarify this; perhaps we can add an "examples" section showing some variations of valid tags (with/without a hostname), wdyt?

File: man/docker-build.1.md [hunk 5]
```diff
@@ -91,7 +91,9 @@ set as the **URL**, the repository is cloned locally and then sent as the contex
    Remove intermediate containers after a successful build. The default is *true*.
 
 **-t**, **--tag**=""
-   Repository names (and optionally with tags) to be applied to the resulting image in case of success.
+   Repository names (and optionally with tags) to be applied to the resulting 
+image in case of success. Refer to **docker-tag(1)** for more information about
+valid tag names.
```
Comment (thaJeztah) [inline man/docker-build.1.md]: nit: can you align these lines with the first one?

File: docs/reference/commandline/tag.md (hunk 4 above)
Comment (thaJeztah) [inline docs/reference/commandline/tag.md]: looks like there are some examples in the man/docker-tag.1.md file, we should have the same information in the online documentation and the man pages (examples there may need a refresh though)

File: man/docker-tag.1.md [hunk 6]
```diff
@@ -34,7 +34,8 @@ separated by a ':'
    The tag you are assigning to the image.  Though this is arbitrary it is
 recommended to be used for a version to distinguish images with the same name.
 Also, for consistency tags should only include a-z0-9-_. .
-Note that here TAG is a part of the overall name or "tag".
+Note that here TAG is a part of the overall name or "tag". Refer to 
+**docker-tag(1)** for more information about valid image and tag names.
```
Comment (thaJeztah) [inline man/docker-tag.1.md]: This is the "docker-tag(1)" man-page, so we should copy the information here that's in the docs/reference/commandline/tag.md document.

**Comment by thaJeztah** [comment]
Thanks @isubuz I left some notes / suggestions. Also, can you squash your commits? I think it's ok to have all changes in a single commit for this PR.

**Comment by isubuz** [comment]
@thaJeztah sure, will do so. Will add the required examples.

**Comment by isubuz** [comment]
Pushed latest changes. Added few more examples for `docker tag`.

**Comment by isubuz** [comment]
ping... 🙂

**Comment by thaJeztah** [comment]
oh! sorry, thanks for the ping. LGTM!

**Comment by thaJeztah** [comment]
think we have enough LGTM's now, so I'll merge this 🎉

**Comment by isubuz** [comment]
Thanks! First contrib to docker! 🎉 🎉 🎉

**Comment by thaJeztah** [comment]
Congrats @isubuz, and thanks so much for contributing!
